### PR TITLE
Update  --select arg to accept specific rules

### PR DESCRIPTION
**Pull request by soulitzer**

### CODE REVIEW

File: torchfix/torchfix.py [hunk 1]
```diff
@@ -25,18 +25,56 @@
...
+        TorchVisionDeprecatedPretrainedVisitor,
+        TorchVisionDeprecatedToTensorVisitor,
+        TorchUnsafeLoadVisitor,
+        TorchReentrantCheckpointVisitor,
     ]
 
+_ALL_ERROR_CODES = None
+
+def GET_ALL_ERROR_CODES():
+    global _ALL_ERROR_CODES
```
Comment (kit1980) [inline torchfix/torchfix.py:42]: Is global needed here?

File: torchfix/torchfix.py [hunk 2]
```diff
@@ -25,18 +25,56 @@
 
 DISABLED_BY_DEFAULT = ["TOR3", "TOR4"]
 
-
-def GET_ALL_VISITORS():
+def GET_ALL_VISITOR_CLS():
```
Comment (kit1980) [inline torchfix/torchfix.py:28]: As we don't call the visitors here anymore, should be just 

ALL_VISITORS_CLS = [
        TorchDeprecatedSymbolsVisitor,
        TorchRequireGradVisitor,
        TorchSynchronizedDataLoaderVisitor,
        TorchVisionDeprecatedPretrainedVisitor,
        TorchVisionDeprecatedToTensorVisitor,
        TorchUnsafeLoadVisitor,
        TorchReentrantCheckpointVisitor,
]

File: torchfix/torchfix.py (hunk 1 above)
Comment (soulitzer) [reply torchfix/torchfix.py:42]: oh whoops, I was supposed to set it...

**Comment by kit1980** [comment]
@soulitzer could you resolve the conflict so CI can run?

**Comment by soulitzer** [comment]
Updated!

**Comment by kit1980** [comment]
Now the tests are failing because https://github.com/pytorch-labs/torchfix/blob/main/tests/test_torchfix.py#L23 passes `select="ALL"`, but now it's `select: List[str]`.

Also `select: List[str] = None` will not pass a typechecker.

File: torchfix/torchfix.py [hunk 3]
```diff
@@ -25,18 +25,55 @@
...
+    TorchVisionDeprecatedPretrainedVisitor,
+    TorchVisionDeprecatedToTensorVisitor,
+    TorchUnsafeLoadVisitor,
+    TorchReentrantCheckpointVisitor,
+]
+
+_ALL_ERROR_CODES = None
+
+def GET_ALL_ERROR_CODES():
+    global _ALL_ERROR_CODES
```
Comment (kit1980) [inline torchfix/torchfix.py:42]: Here `_ALL_ERROR_CODES` just to cache and don't recompute?
Maybe just use `@functools.cache` then?
(It's ok to refactor in later in a different PR).

File: torchfix/torchfix.py [hunk 4]
```diff
@@ -25,18 +25,55 @@
...
-
+    out = []
+    for v in ALL_VISITOR_CLS:
+        if v is TorchDeprecatedSymbolsVisitor:
+            out.append(v(DEPRECATED_CONFIG_PATH))
+        else:
+            out.append(v())
+    return out
+
+def get_visitor_with_error_code(error_code):
```
Comment (kit1980) [inline torchfix/torchfix.py:64]: The API (same for ruff and flake8) is that if, for example, `select=TOR0`, then TOR001, TOR002, TOR003 are enabled.
So we should check for `startswith`, not equality.
And multiple visitors can correspond to the same string in `select`.

File: torchfix/torchfix.py (hunk 3 above)
Comment (soulitzer) [reply torchfix/torchfix.py:42]: Good point, will update here

File: torchfix/torchfix.py [hunk 5]
```diff
@@ -25,18 +25,55 @@
...
+                return visitor
+        else:
+            if error_code == visitor.ERROR_CODE:
+                return visitor
+    assert False, f"Unknown error code: {error_code}"
+
+def get_visitors_with_error_codes(error_codes):
+    visitors = []
+    for error_code in error_codes:
+        visitors.append(get_visitor_with_error_code(error_code))
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torchfix/torchfix.py]: Good catch, updated.

**Comment by soulitzer** [comment]
All the "ALL", default exclude, expansion handling is done in the arg processing/validation stage, so we don't need to deal with those later.

**Comment by kit1980** [comment]
Looks good overall, but there are couple of bugs:

```
$ torchfix . --select=ALL
Traceback (most recent call last):
  File "/home/sdym/.conda/envs/py39/bin/torchfix", line 8, in <module>
    sys.exit(main())
  File "/home/sdym/.conda/envs/py39/lib/python3.9/site-packages/torchfix/__main__.py", line 128, in main
    config.select = process_error_code_str(args.select)
  File "/home/sdym/.conda/envs/py39/lib/python3.9/site-packages/torchfix/__main__.py", line 34, in process_error_code_str
    raise ValueError(f"Invalid error code: {c}, available error "
ValueError: Invalid error code: ALL, available error codes: ['TOR003', 'TOR102', 'TOR002', 'TOR201', 'TOR202', 'TOR001', 'TOR101', 'TOR401']
```

```
$ torchfix . --select=TOR001
Traceback (most recent call last):
  File "/home/sdym/.conda/envs/py39/bin/torchfix", line 8, in <module>
    sys.exit(main())
  File "/home/sdym/.conda/envs/py39/lib/python3.9/site-packages/torchfix/__main__.py", line 128, in main
    config.select = process_error_code_str(args.select)
  File "/home/sdym/.conda/envs/py39/lib/python3.9/site-packages/torchfix/__main__.py", line 40, in process_error_code_str
    return list(expand_error_codes(raw_codes))
TypeError: unhashable type: 'list'
```

**Comment by soulitzer** [comment]
Damn maybe I should've ran the code wtf lol

**Review by kit1980 — APPROVED** [review]

**Comment by kit1980** [comment]
Thanks!

**Pull request merged**